File: Flip Equivalent Binary Trees.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    bool flipEquiv(TreeNode* root1, TreeNode* root2) {
        function<bool(TreeNode*, TreeNode*)> recur;
        recur = [&recur] (TreeNode* node1, TreeNode* node2) -> bool {
            if (node1 == nullptr && node2 == nullptr) return true;
            if (node1 == nullptr || node2 == nullptr) return false;
            if (node1->val != node2->val) return false;
            if (!node1->left && !node2->right) return recur(node1->right, node2->left);
            if (!node1->right && !node2->left) return recur(node1->left, node2->right);
            // cout << node2->left << endl;
            if (node1->left && node2->left && node1->left->val != node2->left->val) return recur(node1->left, node2->right) && recur(node1->right, node2->left);
            return recur(node1->left, node2->left) && recur(node1->right, node2->right);
        };
        return recur(root1, root2);
    }
};

File: Flip Equivalent Binary Trees.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func flipEquiv(root1 *TreeNode, root2 *TreeNode) bool {
    var recur func(node1, node2 *TreeNode) bool
    recur = func(node1, node2 *TreeNode) bool {
        switch {
            case node1 == nil && node2 == nil:
            return true
            case node1 == nil || node2 == nil:
            return false
            case node1.Val != node2.Val:
            return false
            case node1.Left == nil && node2.Right == nil:
            return recur(node1.Right, node2.Left)
            case node1.Right == nil && node2.Left == nil:
            return recur(node1.Left, node2.Right)
            case node1.Left != nil && node2.Left != nil && node1.Left.Val != node2.Left.Val:
            return recur(node1.Left, node2.Right) && recur(node1.Right, node2.Left)
            default:
            return recur(node1.Left, node2.Left) && recur(node1.Right, node2.Right)
        }
    }
    return recur(root1, root2)
}
